sql: move schema change backfill logic to sql/backfill

vivekmenezes opened the pull request:
The distsql backfill calls the sql/backfill code.

The future plan is to also call backfill within a transaction
that runs schema changes in the same transaction as a
CREATE TABLE. Since the CREATE TABLE in a transaction
is not visible to other transactions until the transaction
commits, the schema change can run the backfill without
incrementing the version of the table descriptor.

related to #24626

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25362)
<!-- Reviewable:end -->

vivekmenezes commented:
it's really a refactor to allow use of distsql processor code on a local execution path. One option is to  create sql/processors and put it there?

mjibson commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/sqlbase/backfill.go, line 377 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25362#-LCUbCMy841PTcCwnv6Z:-LCUbCMzAA_sZg81f-ue:btq7w5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8488b333c021829d52c7190ba430131b4e0427c/pkg/sql/sqlbase/backfill.go#L377)):*
> ```Go
> 		encRow, _, _, err := ib.fetcher.NextRow(ctx)
> 		if err != nil {
> 			err = scrub.UnwrapScrubError(err)
> ```

To address andrei's comment: I think we can just outright remove this check, and thus the package import. It was included in all calls of the above function (fetcher.NextRow) but we know that we are never doing anything scrub related in the backfiller, and thus can never produce that error.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25362)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: move schema change backfill logic to sql/backfill

The distsql backfill calls the sql/backfill code.

The future plan is to also call backfill within a transaction
that runs schema changes in the same transaction as a
CREATE TABLE. Since the CREATE TABLE in a transaction
is not visible to other transactions until the transaction
commits, the schema change can run the backfill without
incrementing the version of the table descriptor and running
the schema change state machine.

related to #24626

Release note: None

vivekmenezes commented:
I've moved the common backfill code to sql/backfill and have gotten rid of the scrub dependency.

vivekmenezes commented:
bors r+
